Automated Load-Store-Queue Sizing

patschu11 opened the pull request:
This PR adds a pass, which sizes the Load and Store Queues, to the minimal depth, which still allows optimal throughput.
This is done by analyzing the timing of LSQ loads/stores for each CFDFC separately and using the maximum queue sizes of any CFDFC. The pass is run between the handshake-canonicalization and the handshake-hw conversion. 

### Important side notes
1. There were wrong latencies for some components in the components.json file, I fixed them to match the latencies I observed in the simulation.
2. Currently this algorithm only works with *fpga20-legacy* buffer placement, since in *fpl22* and simple-buffer placement, not all the information is passed through the attributes. When not all necessary information is available, and therefore LSQ sizing cannot be done, the default depth will be assigned to the LSQs.

## Overview
The following things were implemented:
1. The pass receives information on CFDFCs and their respective throughputs from the buffer-placement pass via Handshake attributes (see #105 ). 
1. With this information the CFDFCs are reconstructed and the sizing algorithm determines the queue depths according to the following paper: [ Load-Store Queue Sizing for
Efficient Dataflow Circuits](https://dynamo.ethz.ch/wp-content/uploads/sites/22/2023/01/Load-Store_Queue_Sizing_for_Efficient_Dataflow_Circuits.pdf) 
1. The sizing information is passed on as a Handshake attribute in the corresponding LSQ operation. The attribute is a tuple, with first the Load-Queue size and second the Store-Queue size. e.g. ``` lsqDepth = #handshake<lsqDepth[3, 5]> ```
1. Furthermore the logic to read this information is added in the legacy backend and the new backend to pass it on to the LSQ generator. For this, the ```LSQGenerationInfo``` class has been modified to read the ```lsqDepth``` attribute. If the attribute does not exist, the default size (16) will be used for both the load and store queue.

## Arguments
The pass needs the compontents.json file as a timingDB, which is passed similarly to buffer placement. Furthermore, it is possible to specify different collision scenarios to the pass, by the ```collisions=none/half/full``` argument. These assume one of the following cases.
1. *none*: None of the LSQ memory operations have collisions
1. *half*: Every other LSQ memory operation has collisions
1. *full* All of the LSQ memory operations have collisions

## Details
### Algorithm data structure
The sizing algorithm works on the different CFDFCs of the kernel, determined in buffer placement. The CFDFCs are represented, as an adjacency list (AdjListGraph made of AdjListNodes), since most of the algorithm steps are variations of path-finding between different nodes.


## Detailed steps of the algorithm
The sizing algorithm inside the LSQ Sizing pass has the following steps:
1. Reconstuct the CFDFCs
1. Identifying Start Node and Phi Nodes
1. Adding additional edges to account for some edge cases
1. Find the earliest start time of each node and save it 
1. Finding the worst-case II for each CFDFC
1. Set correct IIs for different collision scenarios for each CFDFC
1. Determine allocation times for each CFDFC
1. Determine deallocation times for each CFDFC
1. Calculate queue depths based on all the timing information and IIs from all CFDFCs

The implementation details are described in the semester thesis report (*Automated Load-Store Queue Sizing in Dynamatic based on MLIR*)

## Results
The results match the one of the paper, namely the algorithm can find the optimal LSQ size for the steady-state, given that the correct sequence of IIs is provided, either by buffer placement or manually. The detailed results can be found in the semester thesis report: (*Automated Load-Store Queue Sizing in Dynamatic based on MLIR*)

Jiahui17 commented:
@patschu11 Hi Patrick thank you so much for the contribution!

Before digging into the detailed stuff (i.e., coding style, naming, etc), I have a very high-level comment on the PR: 
- Why is the LSQ sizing a separate pass instead of inside the buffer placement pass (e.g., after instantiating all buffers in `placeUsingMILP`)? The buffer pass seems to have most of the functionalities that you have re-implemented in the PR (e.g., units and channels inside a CFDFC, etc.).
https://github.com/EPFL-LAP/dynamatic/blob/709f10362d2d5887042f2a8a48bc0cd8d064a8e5/lib/Transforms/BufferPlacement/HandshakePlaceBuffers.cpp#L194-L198

patschu11 commented:
> @patschu11 Hi Patrick thank you so much for the contribution!
> 
> Before digging into the detailed stuff (i.e., coding style, naming, etc), I have a very high-level comment on the PR:
> 
> * Why is the LSQ sizing a separate pass instead of inside the buffer placement pass (e.g., after instantiating all buffers in `placeUsingMILP`)? The buffer pass seems to have most of the functionalities that you have re-implemented in the PR (e.g., units and channels inside a CFDFC, etc.).
>   https://github.com/EPFL-LAP/dynamatic/blob/709f10362d2d5887042f2a8a48bc0cd8d064a8e5/lib/Transforms/BufferPlacement/HandshakePlaceBuffers.cpp#L194-L198

Hi Jiahui,

Although on a high level some functionalities that the buffer placement implements are similar, the data structures (e.g. CFDFC class) and corresponding algorithms are not sufficient for LSQ sizing. Only knowing the units and channels of a CFDFC would not have been enough, since there are some special features needed  (e.g. inserting pseudo nodes into the graph that do not correspond to any Operation). Therefore it was easier and cleaner to implement the LSQ sizing creating a dedicated data structure for the graph instead of the CFDFC stuff from buffer placement. Implementing these features in the existing buffer placement structures would have necessitated a lot of changes, which could have impacted other features and the people working on them.
We discussed this in the beginning with Lucas and decided to make it a dedicated pass since barely any functionality or information could be re-used from buffer placement. Then it is also possible for the user to decide if they want to use the LSQ sizing pass or tune the size manually instead. Also, the list of BBs and II for every CFDFC will be needed by other passes in the future as well, as far as I know, therefore it made sense to us, to encode this information as attributes, instead of building a wrapper around buffer placement.

lucas-rami commented:
Thanks for the contribution and for tagging me on this. I will try to take a look at this early this week, probably focusing more on the parts of the "regular flow" that are affected, as opposed to the new pass itself.

lucas-rami commented on tools/dynamatic/scripts/compile.sh:
I don't think the compile script should be modified at this point since this is an experimental pass (which also "downgrades" the buffer placement algorithm we can use compared to our best algorithm).

patschu11 replied:
Should I add the pass in a way that it is only used with the experimental flag or don't change the compile.sh at all?

lucas-rami replied:
I would recommend not modifying the compile script at all at this stage. I don't know how much this has been tested so I just want to avoid seeing crashes on the default flow. It would be nice to add some unit tests for the pass in a future contribution, then we can consider moving this to non-experimental and to the default compilation flow at the same time.

patschu11 replied:
Ok, I completely removed LSQ sizing from the compile script now.

lucas-rami commented on experimental/include/experimental/Transforms/LSQSizing/HandshakeSizeLSQs.h:
```diff
@@ -0,0 +1,30 @@
+} // namespace experimental
+} // namespace dynamatic
+
+#endif // DYNAMATIC_TRANSFORMS_SIZE_LSQS_H
```
nit pick: add empty line at the end of the file

lucas-rami commented on experimental/include/experimental/Transforms/LSQSizing/HandshakeSizeLSQs.h:
```diff
@@ -0,0 +1,30 @@
+//===- HandshakeSizeLSQs.h - Sizes the LSQs --------*- C++ -*-===//
```
Insert our standard copyright notice and make the first line 80-characters long. Please make sure this is the case in all new files you added.

lucas-rami commented on experimental/include/experimental/Transforms/LSQSizing/LSQSizingSupport.h:
```diff
@@ -0,0 +1,164 @@
+} // namespace experimental
+} // namespace dynamatic
+
+#endif // DYNAMATIC_TRANSFORMS_LSQSIZING_LSQSIZINGSUPPORT_H
```
nit pick: add empty line at the end of the file

lucas-rami commented on experimental/include/experimental/Transforms/LSQSizing/LSQSizingSupport.h:
```diff
@@ -0,0 +1,164 @@
+namespace experimental {
+namespace lsqsizing {
+
+// Define a structure for a graph node
```
By convention all entity comments (e.g., struct, class, functions, members, etc.) should start with `///` instead of `//`. This is so that automated documentation parsing tools can recognize them more easily (I think). Please try to make a quick pass over your files and use `///` in those scenarios.

lucas-rami commented on experimental/include/experimental/Transforms/LSQSizing/LSQSizingSupport.h:
```diff
@@ -0,0 +1,164 @@
+
+// Define a structure for a graph node
+struct AdjListNode {
+  int latency;                 // Latency of the operation
```
Do not put in-line comments on lines which also contain code, put the comment above instead.

```cpp
/// Latency of the operation.
int latency;
```

The former has two disadvantaged:

1. It tends to create weird indentation with clang-format when wrapping a comment over multiple lines, hurting readability.
2. When commenting type members like here, I don't think most IDEs will pick up that the comment is associated to the member; users will not see the documentation when hovering over the member.

lucas-rami commented on experimental/include/experimental/Transforms/LSQSizing/LSQSizingSupport.h:
```diff
@@ -0,0 +1,164 @@
+  std::set<std::string> edges; // Adjacency list (stores keys of adjacent nodes)
+  std::set<std::string> backedges; // Backedge list (stores keys of adjacent
+                                   // nodes connected by backedges)
+  std::set<std::string>
+      shiftingedges; // Shifting edge list (stores keys of adjacent nodes
+                     // connected by shifting edges)
```
A "weird indentation" example (related to my previous comment).

lucas-rami commented on experimental/include/experimental/Transforms/Passes.td:
```diff
@@ -97,4 +97,20 @@ def CreditBasedSharing : DynamaticPass<"credit-based-sharing"> {
+  let description = [{
+    Sizes the Load-Store-Queues based on the buffer placement information to ensure
+    that the memory accesses are not restricting the circuits troughput, while 
+    trying to keep the area as low as possible. TODO  
```
Either remove TODO or complete the description.

lucas-rami commented on experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp:
```diff
@@ -0,0 +1,672 @@
+  // allocation/deallocation time inside the LSQ For allocation it is the same
+  // for both load and stores, for deallocation it is different between loads
+  // and stores
+  static const int allocEntryLatency = 1;
+  static const int storeDeallocEntryLatency = 2;
+  static const int loadDeallocEntryLatency = 1;
```
Can latencies be negative? If not use `unsigned` instead of `int`.

lucas-rami commented on experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp:
```diff
@@ -0,0 +1,672 @@
+    handshake::CFDFCToBBListAttr cfdfcAttr =
+        getDialectAttr<handshake::CFDFCToBBListAttr>(funcOp);
+
+    if (throughputAttr == nullptr || cfdfcAttr == nullptr) {
```
nit pick: no braces around single-statement bodies ([LLVM coding guidelines](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements)). Applies in other places as well.

lucas-rami commented on experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp:
```diff
@@ -0,0 +1,672 @@
+    std::unordered_map<unsigned, float> IIs;
+
+    // Extract CFDFCs and II from the Attributes
+    handshake::CFDFCThroughputAttr throughputAttr =
```
nit pick: since the return type is also indicated on the right-hand side of the assignment you can use `auto` to be a bit more concise here ([LLVM coding guidelines](https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable)).

lucas-rami commented on experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp:
```diff
@@ -0,0 +1,672 @@
+    DictionaryAttr cfdfcDict = cfdfcAttr.getCfdfcMap();
+
+    // Convert Attribute into better usable data structure
+    for (auto &attr : cfdfcDict) {
```
Avoid using `auto` when it's not obvious just by looking at the line what the iterator type will be and the type is "not horrible to type" (loose guideline, as you can imagine). Here it is just `NamedAttribute&` and I think it improves readability to write it out fully.

lucas-rami commented on experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp:
```diff
@@ -0,0 +1,672 @@
+    for (auto &attr : cfdfcDict) {
+      ArrayAttr bbList = llvm::dyn_cast<ArrayAttr>(attr.getValue());
+      llvm::SetVector<unsigned> cfdfcBBs;
+      for (auto &bb : bbList) {
```
No need to pass the `Attribute` by reference. An attribute is basically a fancy pointer itself so they are usually passed by value (see `Attribute`'s documentation for details).

lucas-rami commented on experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp:
```diff
@@ -0,0 +1,672 @@
+      for (auto &bb : bbList) {
+        cfdfcBBs.insert(bb.cast<IntegerAttr>().getUInt());
+      }
+      cfdfcBBLists.insert({std::stoi(attr.getName().str()), cfdfcBBs});
```
Use `StringRef::getAsInteger` method instead of `std::stoi` to gracefully catch eventual parsing errors (`std::stoi` will just throw an exception in case of error, crashing the program). Here that would probably look like

```cpp
unsigned index;
if (attr.getName().getAsInteger(10, index)) {
  // report an error
}
// everything is fine, you can use index
```

patschu11 replied:
What would be the best way to report the error?
1. printing something with `llvm::errs()`, skipping the CFDFC and then continue 
2. calling `signalPassFailure()`
3. something else?

lucas-rami replied:
Since you are in `runDynamaticPass` doing `return signalPassFailure();` seems like the best option here.

lucas-rami commented on experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp:
```diff
@@ -0,0 +1,672 @@
+  // Find the node with the highest latency, if there are multiple, choose the
+  // one with the most nodes
+  mlir::Operation *maxLatencyNode = nullptr;
+  int maxLatency = 0;
+  int maxNodeCount = 0;
```
Use `unsigned` if these numbers cannot be negative.

lucas-rami commented on experimental/lib/Transforms/LSQSizing/LSQSizingSupport.cpp:
```diff
@@ -0,0 +1,574 @@
+// 2. If the operation is a buffer operation, the latency is extracted from the
+// timing attribute
+// 3. If the operation is neither, then its latency is set to 0
+int extractNodeLatency(mlir::Operation *op, TimingDatabase timingDB) {
```
Should be `static`.

lucas-rami commented on experimental/lib/Transforms/LSQSizing/LSQSizingSupport.cpp:
```diff
@@ -0,0 +1,574 @@
+  }
+
+  // For each LSQ, go trough all loads and find the maxPathLatency to all stores
+  int maxLatency = 0;
```
Use `unsigned`.

lucas-rami commented on experimental/lib/Transforms/LSQSizing/LSQSizingSupport.cpp:
```diff
@@ -0,0 +1,574 @@
+  } else {
+    return nodes.at(opName).earliestStartTime;
+  }
+}
```
nit: add empty line at the end of the file

lucas-rami commented on experimental/lib/Transforms/LSQSizing/LSQSizingSupport.cpp:
```diff
@@ -0,0 +1,574 @@
+  for (const auto &pair : nodes) {
+    std::string opName = pair.first;
+    const AdjListNode &node = pair.second;
+    llvm::dbgs() << opName << " (lat: " << node.latency
```
Unless you have a good reason, you should use `llvm::errs()` for this kind of print statements. AFAIK `llvm::dbgs()` is more supposed to work in tandem with `LLVM_DEBUG` macros and such.

lucas-rami commented on experimental/lib/Transforms/LSQSizing/LSQSizingSupport.cpp:
```diff
@@ -0,0 +1,574 @@
+  // adding shifting edges for their start times and should be accounted for
+  if (nodes.at(opName).earliestStartTime == -1) {
+    return 0;
+  } else {
```
No `else` after `return`.

lucas-rami commented on include/dynamatic/Transforms/Passes.td:
```diff
@@ -295,6 +295,7 @@ def HandshakePlaceBuffers : DynamaticPass<"handshake-place-buffers"> {
     "'buffer-placement' at the binary's location.">];
 }
 
+
```
spurious empty line addition

lucas-rami commented on lib/Dialect/Handshake/MemoryInterfaces.cpp:
```diff
@@ -364,6 +364,19 @@ void LSQGenerationInfo::fromPorts(FuncMemoryPorts &ports) {
+    depthLoad = lsqDepthAttr.getLoadQueueDepth();
+    depthStore = lsqDepthAttr.getStoreQueueDepth();
+    depth = std::max(depthLoad,
+                     depthStore); // Parameter is theoretically unused, but
```
Put the comment above the line to not create weird formatting.

lucas-rami reviewed (commented):
Thanks for the big contribution! I left a bunch of remarks, mostly on code style and conventions. Please try to make a quick pass over the entire code with those comments in mind to address them globally (most of the time I only indicated any given "stylistic issue" once, but they most likely apply in many places).

As I said above I did not focus at all on the algorithm you are implementing, someone more informed than me should probably take care of checking the actual thing in detail before this gets merged.

lucas-rami commented:
This is good to merge for me assuming someone else looked at the algorithm being implemented. @Jiahui17 did you or someone else had a look? If so feel free to approve and merge, and thanks for the contribution @patschu11!

Jiahui17 commented on experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp:
```diff
@@ -0,0 +1,661 @@
+      }
+    }
+
+    // Set the maximum Queue sizes as attributes for backend
+    for (auto &maxLoadStoreSize : maxLoadStoreSizes) {
+      mlir::Operation *lsqOp = maxLoadStoreSize.first;
+      unsigned maxLoadSize = std::get<0>(maxLoadStoreSize.second);
+      unsigned maxStoreSize = std::get<1>(maxLoadStoreSize.second);
```
Here (i think) you can do something like this instead:

```c++
for (auto [lsqOp, [maxLoadSize, maxStoreSize]] : maxLoadStoreSizes) {
...
}
```

lucas-rami replied:
I don't think C++ supports nested unpacking, you may have to do

```cpp
for (auto &[lsqOp, maxSizes] : maxLoadStoreSizes) {
  unsigned maxLoadSize = std::get<0>(maxSizes.second);
  unsigned maxStoreSize = std::get<1>(maxSizes.second);
}
```

Jiahui17 commented on experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp:
```diff
@@ -0,0 +1,661 @@
+        sizingResults.push_back(result.value());
+    }
+
+    // Extract maximum Queue sizes for each LSQ
```
There is probably a way to merge this loop with the loop below

Jiahui17 commented:
Hi, Patrick! Thanks again for the contribution

I added a few more comments/questions

lucas-rami commented on experimental/lib/Transforms/LSQSizing/LSQSizingSupport.cpp:
```diff
@@ -0,0 +1,566 @@
+  // other nodes need all of their inputs to be ready
+  // Therefore the latency for mux, merge and cmerge is the minimum of all
+  // incoming edges
+  if (op && (op->getName().getStringRef() == "handshake.mux" ||
+             op->getName().getStringRef() == "handshake.merge" ||
+             op->getName().getStringRef() == "handshake.control_merge")) {
```
Yep never use an operation's name as a check for its nature as it is slower and more error-prone that using its type. Always use `isa`, `cast`, `dyn_cast`, or their few variants. [Here](https://llvm.org/docs/ProgrammersManual.html#isa) is a quick description of these functions.

lucas-rami commented on experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp:
```diff
@@ -0,0 +1,661 @@
+  StartTimes startTimes;
+
+  // Find all mux and control_merge ops as candidates for start node
+  std::vector<mlir::Operation *> muxOps =
```
Yep, even more "LLVM-style".

```cpp
std::vector<mlir::Operation *> startNodeCandidates;
llvm::copy(funcOp.getOps<handshake::MergeLikeOpInterface>(), std::back_inserter(startNodeCandidates);
```

patschu11 replied:
Here again, it is a problem that I only want the Operations of these types inside the specific CFDFC and not of the whole FuncOp. But I created a templated function for my graph, so now it would look like this:
 `  std::vector<mlir::Operation *> startNodeCandidates =
      graph.getOperationsWithOpType<handshake::MergeLikeOpInterface>();`
I think this should also be fine.

patschu11 commented on experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp:
```diff
@@ -0,0 +1,661 @@
+
+  AdjListGraph graph(funcOp, cfdfcBBs, timingDB, initialII);
+  std::vector<mlir::Operation *> loadOps =
+      graph.getOperationsWithOpName("handshake.lsq_load");
```
Here I just want to get operations that are in the CFDFC, not in the whole funcOp, but I changed the function to use the `isa<OpType>(op)` check and made it a template function, so I think that should be quite clean as well.

patschu11 commented on data/components.json:
```diff
@@ -391,7 +391,7 @@
   "handshake.lsq_load": {
     "latency": {
-      "64": 5.0
+      "64": 4.0
```
![image](https://github.com/user-attachments/assets/4bf0bbd3-2799-4d8e-9448-aacde93dbce2)
When looking at the waveforms, I observe a latency of 4. Or am I counting this wrong?

Jiahui17 replied:
So here the point is the minimum possible latency for loading a value from the LSQ is 4? If yes this change makes sense to me :)

patschu11 commented on data/components.json:
```diff
@@ -635,7 +635,7 @@
   "handshake.subf": {
     "latency": {
-      "64": 10.0
+      "64": 8.0
```
I now checked the waveforms again and yes 9 should be correct. I probably just counted wrong back then.

patschu11 commented on experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp:
```diff
@@ -0,0 +1,661 @@
+  // candidate and that the start time of the candidate is not zero
+  insertStartnodeShiftingEdges(graph, startNode, startTimes);
+
+  // Calculate the earliest start times for each operation in the graph. This is
+  // needed to find the allocation times
+  graph.setEarliestStartTimes(startNode);
+
+  std::vector<unsigned> IIs;
+
+  // The collisions argument can be passed in the compile script
+  // For now there are 3 defined cases: full, half, none
+  // full: determines the worst case II, for the case that all lsq ops have
+  // memory collisions
+  // half: determines the worst case II and alternate it with
+  // the best case II (from buffer placement), for the case that half of the lsq
+  // ops have memory collisions
+  // none: gets the best case II(from buffer
+  // placement), for the case that no lsq ops have memory collisions
+  //
+  // It would also easily be possible to add more cases, by just pushing the IIs
+  // into the IIs vector in the order they are occuring
+
+  if (collisions == "full") {
+    IIs.push_back(graph.getWorstCaseII());
+  } else if (collisions == "half") {
+    IIs.push_back(initialII);
+    IIs.push_back(graph.getWorstCaseII());
+  } else {
+    // none is the default case
+    IIs.push_back(initialII);
+  }
```
I don't get how ASAP or SDC-based scheduling relates to this. The best-case II is passed from buffer-placement, I don't know how buffer placement does it. But in my experience, this II was not consistent with the one observed in the simulation (even without any memory dependencies). However, I also don't know if the one from buffer placement is the optimal one. I can just add the note though.

Jiahui17 replied:
> The best-case II is passed from buffer-placement,
> But in my experience, this II was not consistent with the one observed in the simulation (even without any memory dependencies).

Ok good to know. Maybe also leave a note here on the inconsistency between the simulation II vs. the II obtained in MILP. 

What if you calculate the best case II the way you did for the worst case? How does it compare with the simulation II?

patschu11 replied:
I added a comment in the code regarding the inconsistency. 

I did not try that. The calculation for the worst-case II is straightforward, but I am not sure how that would work for the best-case II. Unfortunately, I currently also don't have the time to look further into this. If it would be very important I could potentially check it out in late January/February.

patschu11 commented:
> Thanks again for addressing the comments! Could you also resolve the merge conflict (there is only one loadOp now instead of two separate ones for LSQs and MCs)? There was a big change in #193

Since there is now the same loadOp for LSQ and MC loads, how does this account for them having different latencies? Previously, LsqLoadOp had a latency of 4, whereas the MCLoad had a latency of 1. Now both have a latency of 1, which I don't think is correct. How do you account for this e.g. in buffer placement?

Jiahui17 commented:
> Since there is now the same loadOp for LSQ and MC loads, how does this account for them having different latencies? Previously, LsqLoadOp had a latency of 4, whereas the MCLoad had a latency of 1. Now both have a latency of 1, which I don't think is correct. How do you account for this e.g. in buffer placement?

Thanks for the catch!

For now, we will check if the load is connected to a LSQ when determining the latency of the op: https://github.com/EPFL-LAP/dynamatic/pull/193#issuecomment-2496109013

patschu11 commented:
> > Since there is now the same loadOp for LSQ and MC loads, how does this account for them having different latencies? Previously, LsqLoadOp had a latency of 4, whereas the MCLoad had a latency of 1. Now both have a latency of 1, which I don't think is correct. How do you account for this e.g. in buffer placement?
> 
> Thanks for the catch!
> 
> For now, we will check if the load is connected to a LSQ when determining the latency of the op: [#193 (comment)](https://github.com/EPFL-LAP/dynamatic/pull/193#issuecomment-2496109013)

I have adapted the code to work with this new change. For the LSQLoad latency, I added a check that adds +3 latency (hardcoded), as suggested in the comment from the other pull request. I hope we can finally merge this PR now. Sorry for the long turnaround times on my side.

Jiahui17 commented on experimental/lib/Transforms/LSQSizing/LSQSizingSupport.cpp:
```diff
@@ -559,3 +571,12 @@ int CFDFCGraph::getEarliestStartTime(mlir::Operation *op) {
 }
+
+bool CFDFCGraph::isConnectedToLSQ(mlir::Operation *op) {
+  for (mlir::Operation *destOp : op->getUsers()) {
+    if (isa<handshake::LSQOp>(destOp)) {
+      return true;
+    }
+  }
```
To be safe, use the following to check if the load/store is connected to an LSQ:
https://github.com/EPFL-LAP/dynamatic/blob/f258670bbefca8fdaa12c18bff2c19ca2f93dfdc/lib/Support/TimingModels.cpp#L135-L137

Jiahui17 commented on experimental/lib/Transforms/LSQSizing/LSQSizingSupport.cpp:
```diff
@@ -41,8 +41,20 @@ using namespace dynamatic::experimental::lsqsizing;
   double latency = 0;
 
-  if (!failed(timingDB.getLatency(op, SignalType::DATA, latency)))
+  if (!failed(timingDB.getLatency(op, SignalType::DATA, latency))) {
+
+    // TODO This is a temporary hack to fix the latency of the LoadOp when its
+    // connected to an LSQ (Necessary since there is now the same loadop for MC
+    // and LSQ loads). When the TimingDB gets reworked, this can be removed
+    if (isa<handshake::LoadOp>(op)) {
+      for (mlir::Operation *destOp : op->getUsers()) {
+        if (isa<handshake::LSQOp>(destOp)) {
+          return latency + 3;
+        }
+      }
+    }
```
> I have adapted the code to work with this new change. For the LSQLoad latency, I added a check that adds +3 latency (hardcoded), as suggested in the comment from the other pull request.

Sorry! My previous message was not so clear. 

**You don't need to add this latency here!** Lucas already handled the +3 cycle latency in the getLatency method:
https://github.com/EPFL-LAP/dynamatic/blob/f258670bbefca8fdaa12c18bff2c19ca2f93dfdc/lib/Support/TimingModels.cpp#L114-L141

Jiahui17 commented on experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp:
```diff
@@ -233,8 +233,22 @@ std::optional<LSQSizingResult> HandshakeSizeLSQsPass::sizeLSQsForCFDFC(
+
+  // We only want LSQ loads, therefore we need to check if they are connected to
+  // an LSQ
+  auto allLoadOps = graph.getOperationsWithOpType<handshake::LoadOp>();
+  auto allStoreOps = graph.getOperationsWithOpType<handshake::StoreOp>();
+  std::vector<mlir::Operation *> loadOps;
+  std::vector<mlir::Operation *> storeOps;
+  for (auto op : allLoadOps) {
+    if (graph.isConnectedToLSQ(op))
+      loadOps.push_back(op);
+  }
+
+  for (auto op : allStoreOps) {
+    if (graph.isConnectedToLSQ(op))
+      storeOps.push_back(op);
+  }
```
Maybe just do this instead (same thing for the stores)? You don't need `allLoadOps` later
```c++
std::vector<mlir::Operation *> loadOps;
for (auto op : graph.getOperationsWithOpType<handshake::LoadOp>) {
  if (graph.isConnectedToLSQ(op))
    loadOps.push_back(op);
}
```

Jiahui17 reviewed (changes requested):
Thank you again for the fixes! I still have some comments on the code (sorry to be annoying haha)---after addressing them it would be good to merge :D

patschu11 commented:
Thanks for the quick feedback! I fixed these things now.

Jiahui17 commented on experimental/lib/Transforms/LSQSizing/LSQSizingSupport.cpp:
```diff
@@ -573,10 +562,10 @@ int CFDFCGraph::getEarliestStartTime(mlir::Operation *op) {
 bool CFDFCGraph::isConnectedToLSQ(mlir::Operation *op) {
-  for (mlir::Operation *destOp : op->getUsers()) {
-    if (isa<handshake::LSQOp>(destOp)) {
+  if (auto loadOp = dyn_cast<handshake::LoadOp>(op)) {
+    auto memOp = findMemInterface(loadOp.getAddressResult());
+    if (isa_and_present<handshake::LSQOp>(memOp))
```
What about stores?

patschu11 replied:
Thanks for the catch! But I hope that's the last one :D

Jiahui17 reviewed (changes requested):
One extra..:D

Jiahui17 reviewed (approved):
Good job, thank you Patrick!
